api/base/parsers.py

```diff
@@ -15,6 +17,29 @@ class JSONAPIParser(JSONParser):
     media_type = 'application/vnd.api+json'
     renderer_class = JSONAPIRenderer
 
+    def flatten_relationships(self, relationships):
+        """
+        Flattens relationships dictionary which has information needed to create related resource objects.
+
+        Validates that formatting of relationships dictionary is correct.
+        """
+        if not isinstance(relationships, dict):
+                raise ParseError()
+        related_resource = relationships.keys()[0]
+        if not isinstance(relationships[related_resource], dict) or related_resource == 'data':
+            raise ParseError()
+        data = relationships[related_resource].get('data')
+
+        if not data:
+            raise JSONAPIException(source={'pointer': '/relationships/<related_resource_name>/data'}, detail=NO_DATA_ERROR)
```

Comment:
The link is out of date, but Sam has clarified that it was pointing to "the patterns introduced here: https://github.com/samchrisinger/osf.io/blob/feature/api/comp-filters/api/base/exceptions.py "

I'll let Dawn and Sam decide between them whether this is applicable here and worth pursuing. Preference is to adopt a unified, consistent, and maintainable approach for errors going forward.